File: ch_4/arith.pl
% vim: set ft=prolog:

% term/1 succeeds if the input program is syntactically valid. I haven't
% actually used it in the evaluation predicates (since invalid programs will
% fail anyway) but it's here for completeness.
term(0).
term(true).
term(false).
term(succ(X)) :- term(X).
term(pred(X)) :- term(X).
term(iszero(X)) :- term(X).
term(if(X, Y, Z)) :- term(X), term(Y), term(Z).

% ...ditto for values...
value(X) :- nv(X).
value(true).
value(false).

% ...and finally numeric values (to which the values/1 predicate refers).
nv(0).
nv(succ(X)) :- nv(X).

% result/2 is a wrapper predicate around the reduction rules to tell us when
% to stop (i.e. when we reach a value, either directly or via further
% reduction)
%
% The cuts are here because we know these programs will only ever evaluate to
% a single value, so we don't want to offer a choicepoint.
result(X, X) :- value(X), !.
result(X, Z) :- reduce(X, Y), result(Y, Z), !.

% reduce/2 provides single small-step operational semantic rules for reducing
% a given term.

% boolean reductions
reduce(if(true, X, _), X).
reduce(if(false, _, X), X).
reduce(if(X, Y, Z), if(X1, Y, Z)) :- reduce(X, X1).

% arithmetic reductions
reduce(succ(X), succ(X1)) :- reduce(X, X1).
reduce(pred(0), 0).
reduce(pred(succ(X)), X) :- nv(X).
reduce(pred(X), pred(X1)) :- reduce(X, X1).

% numerical predicate reductions
reduce(iszero(0), true).
reduce(iszero(succ(X)), false) :- nv(X).
reduce(iszero(X), iszero(X1)) :- reduce(X, X1).

File: ch_4/ruby/lib/arith.rb
StuckError = Class.new(StandardError)

class Term
  def value?
    false
  end

  def numeric_value?
    false
  end

  def to_s
    self.class.name
  end

  def reduce
    raise NotImplementedError
  end
end

class Value < Term
  def value?
    true
  end

  def reduce
    raise StuckError
  end
end

class NumericValue < Value
  def numeric_value?
    true
  end
end

True  = Class.new(Value)
False = Class.new(Value)
Zero  = Class.new(NumericValue)

class Succ < Term
  attr_reader :term

  def initialize(term)
    @term = term
  end

  def value?
    numeric_value?
  end

  def numeric_value?
    term.numeric_value?
  end

  def to_s
    "Succ(#{term})"
  end

  def reduce
    self.class.new(term.reduce)
  end
end

def Succ(term)
  Succ.new(term)
end

class Pred < Term
  attr_reader :term

  def initialize(term)
    @term = term
  end

  def to_s
    "Pred(#{term})"
  end

  def reduce
    case
    when term.is_a?(Zero) then term
    when term.is_a?(Succ) && term.term.numeric_value? then @term.term
    else Pred.new(term.reduce)
    end
  end
end

def Pred(term)
  Pred.new(term)
end

class IsZero < Term
  attr_reader :term

  def initialize(term)
    @term = term
  end

  def to_s
    "IsZero(#{term})"
  end

  def reduce
    case
    when term.is_a?(Zero) then True.new
    when term.is_a?(Succ) && term.term.numeric_value? then False.new
    else IsZero.new(term.reduce)
    end
  end
end

def IsZero(term)
  IsZero.new(term)
end

class If < Term
  attr_reader :test, :consequent, :alternate

  def initialize(test, consequent, alternate)
    @test, @consequent, @alternate = test, consequent, alternate
  end

  def to_s
    "If(#{test}, #{consequent}, #{alternate})"
  end

  def reduce
    case test
    when True then consequent
    when False then alternate
    else If.new(test.reduce, consequent, alternate)
    end
  end
end

def If(test, consequent, alternate)
  If.new(test, consequent, alternate)
end

class Machine
  def evaluate(term)
    if term.value?
      term
    else
      evaluate(term.reduce)
    end
  end
end

term = If(
    IsZero(Pred(Succ(Pred(Succ(Zero.new))))),
    Succ(Succ(Zero.new)),
    If(False.new, Zero.new, Succ(Zero.new))
  )

puts term
puts "=>"
puts Machine.new.evaluate(term)
